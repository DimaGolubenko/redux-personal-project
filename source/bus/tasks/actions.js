//Types
import { types } from "./types";

export const tasksActions = {
    //Sync
    fillTasks: (tasks) => {
        return {
            type:    types.FILL_TASKS,
            payload: tasks,
        };
    },
    createTask: (task) => {
        return {
            type:    types.CREATE_TASK,
            payload: task,
        };
    },
    removeTask: (taskId) => {
        return {
            type:    types.REMOVE_TASK,
            payload: taskId,
        };
    },
    updateTask: (message) => {
        return {
            type:    types.UPDATE_TASK,
            payload: message,
        };
    },
    updateTasksFilter: (message) => {
        return {
            type:    types.UPDATE_TASKS_FILTER,
            payload: message,
        };
    },
    completeAllTasks: (tasks) => {
        return {
            type:    types.COMPLETE_ALL_TASKS,
            payload: tasks,
        };
    },
    updateNewTaskMessage: (message) => {
        return {
            type:    types.UPDATE_NEW_TASK_MESSAGE,
            payload: message,
        };
    },
    updateEditingTaskMessage: (message) => {
        return {
            type:    types.UPDATE_EDITING_TASK_MESSAGE,
            payload: message,
        };
    },
    clearEditingTaskMessage: () => {
        return {
            type: types.CLEAR_EDITING_TASK_MESSAGE,
        };
    },
    toggleTaskEditing: () => {
        return {
            type: types.TOGGLE_TASK_EDITING,
        };
    },
    setEditingTaskId: (id) => {
        return {
            type:    types.SET_EDITING_TASK_ID,
            payload: id,
        };
    },
    //Async
    fetchTasksAsync: () => {
        return {
            type: types.FETCH_TASKS_ASYNC,
        };
    },
    createTaskAsync: (task) => {
        return {
            type:    types.CREATE_TASK_ASYNC,
            payload: task,
        };
    },
    removeTaskAsync: (taskId) => {
        return {
            type:    types.REMOVE_TASK_ASYNC,
            payload: taskId,
        };
    },
    updateTaskAsync: (task) => {
        return {
            type:    types.UPDATE_TASK_ASYNC,
            payload: task,
        };
    },
    completeAllTasksAsync: (tasks) => {
        return {
            type:    types.COMPLETE_ALL_TASKS_ASYNC,
            payload: tasks,
        };
    },
};
