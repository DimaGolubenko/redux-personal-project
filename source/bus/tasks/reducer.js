//Core
import { fromJS, List, Map } from "immutable";

//Instruments
import { types } from "./types";

const initialState = {
    tasks:              List(),
    tasksFilter:        "",
    newTaskMessage:     "",
    editingTaskMessage: "",
    isTaskEditing:      false,
    editingTaskId:      null,
};

export const taskReducer = (state = initialState, action) => {
    const { payload } = action;

    switch (action.type) {
        case types.FILL_TASKS:
            return {
                ...state,
                tasks: fromJS(payload),
            };
        case types.CREATE_TASK: {
            return {
                ...state,
                tasks: state.tasks.unshift(fromJS(payload)),
            };
        }
        case types.REMOVE_TASK: {
            return {
                ...state,
                tasks: state.tasks.filter((task) => task.get("id") !== payload),
            };
        }
        case types.UPDATE_TASK: {
            const { payload: updatedTask } = action;

            return {
                ...state,
                tasks: state.tasks.map((task) =>
                    task.get("id") === updatedTask.id ? Map(updatedTask) : task
                ),
            };
        }
        case types.UPDATE_TASKS_FILTER: {
            const { payload: tasksFilter } = action;

            return {
                ...state,
                tasksFilter,
            };
        }
        case types.COMPLETE_ALL_TASKS:
            return {
                ...state,
                tasks: fromJS(payload),
            };
        case types.UPDATE_NEW_TASK_MESSAGE:
            return {
                ...state,
                newTaskMessage: payload,
            };
        case types.UPDATE_EDITING_TASK_MESSAGE:
            return {
                ...state,
                editingTaskMessage: payload,
            };
        case types.CLEAR_EDITING_TASK_MESSAGE:
            return {
                ...state,
                editingTaskMessage: "",
            };
        case types.TOGGLE_TASK_EDITING:
            return {
                ...state,
                isTaskEditing: !state.isTaskEditing,
            };
        case types.SET_EDITING_TASK_ID:
            return {
                ...state,
                editingTaskId: payload,
            };
        default:
            return state;
    }
};
