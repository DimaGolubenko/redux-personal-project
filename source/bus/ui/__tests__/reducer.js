//Reducer
import { uiReducer } from "../reducer";

//Actions
import { uiActions } from "../actions";

describe("ui reducer", () => {
    test("should return initial state by default", () => {
        expect(uiReducer(void 0, {})).toMatchSnapshot();
    });

    test("should handle START_FETCHING action", () => {
        expect(
            uiReducer(void 0, uiActions.startFetching(__.ui.isFetching))
        ).toMatchSnapshot();
    });

    test("should handle STOP_FETCHING action", () => {
        expect(
            uiReducer(void 0, uiActions.stopFetching(__.ui.isNotFetching))
        ).toMatchSnapshot();
    });
});
