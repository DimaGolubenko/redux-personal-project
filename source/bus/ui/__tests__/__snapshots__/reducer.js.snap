// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui reducer should handle START_FETCHING action 1`] = `
Immutable.Map {
  "isFetching": true,
}
`;

exports[`ui reducer should handle STOP_FETCHING action 1`] = `
Immutable.Map {
  "isFetching": false,
}
`;

exports[`ui reducer should return initial state by default 1`] = `
Immutable.Map {
  "isFetching": false,
}
`;
